src/migrations/merge.rs

```diff
@@ -0,0 +1,147 @@
+use indexmap::IndexMap;
+use fs_err as fs;
+
+use crate::connect::Connection;
+use crate::migrations::{Context, migrate, migration};
+use crate::migrations::create::{MigrationKey, MigrationToText, write_migration};
+use crate::migrations::db_migration::{DBMigration, read_all};
+use crate::migrations::migration::MigrationFile;
+use crate::migrations::rebase::{fix_migration_ids};
+
+pub struct MergeMigrations {
+    pub base_migrations: IndexMap<String, MergeMigration>,
+    pub target_migrations: IndexMap<String, MergeMigration>
+}
+
+impl MergeMigrations {
+    fn flatten(&self) -> IndexMap<&String, &MergeMigration> {
+        let mut result = IndexMap::new();
+
+        for migration in self.base_migrations.iter() {
+            result.insert(migration.0, migration.1);
+        }
+
+        for migration in self.target_migrations.iter() {
+            result.insert(migration.0, migration.1);
+        }
+
+        return result;
+    }
+}
+
+pub struct MergeMigration {
+    key: MigrationKey,
+    id_override: Option<String>,
+    migration: DBMigration,
+    parent_override: Option<String>
+}
+
+impl MigrationToText for MergeMigration {
+    type StatementsIter<'a> = std::iter::Once<&'a String> where Self: 'a;
+
+    fn key(&self) -> &MigrationKey {
+        &self.key
+    }
+
+    fn parent(&self) -> anyhow::Result<&str> {
+        Ok(self.parent_override.as_ref().or(self.migration.parent_names.first()).map(|v| v.as_str()).unwrap_or("initial"))
+    }
+
+    fn id(&self) -> anyhow::Result<&str> {
+        Ok(self.id_override.as_ref().unwrap_or(&self.migration.name))
+    }
+
+    fn statements<'a>(&'a self) -> Self::StatementsIter<'a> {
+        std::iter::once(&self.migration.script)
+    }
+}
+
+pub async fn get_merge_migrations(base: &mut Connection, target: &mut Connection) -> anyhow::Result<MergeMigrations> {
+    let base_migrations = read_all(base, true, false).await?;
+    let mut target_migrations = read_all(target, true, false).await?;
+
+    if base_migrations.len() > target_migrations.len() {
+        anyhow::bail!("Source branch contains more migrations than the target branch");
+    }
+
+    for (index, (base_migration_id, _)) in base_migrations.iter().enumerate() {
+        if let Some((target_migration_id, _)) = target_migrations.get_index(index) {
+            if target_migration_id != base_migration_id {
+                anyhow::bail!("Migration histories of base and target diverge");
+            }
+        } else {
+            anyhow::bail!("Target branch doesn't contain the migration {} (at index {})", base_migration_id, index)
+        }
+    }
```

Comment:
These error messages do explain what's going on in our merge algorithm, but I don't think people who know git would find them useful.

What I would expect is only the following error message:
```
Cannot fast-forward.
Current branch contains migrations that are not target branch:
- 00003.edgeql
- 00004.edgeql
- 00005.edgeql

Hint: you must rebase the target branch onto current branch first:
    edgedb branch switch {target-branch}
    edgedb branch rebase {current-branch}
    edgedb branch switch {current-branch}
    edgedb branch merge {target-branch}
```

One way to do this is to:
- first find the merge base, same as you have done with rebase.
- check that `merge_base == current_branch`, otherwise throw the error above,
- `target_migrations = merge_base..target_branch`

Note that if there is no common migrations, merge base is `initial`. This means that all migrations on target branch are `target_migrations` and that if there are any migrations on current branch, we throw the error above.

Reply:
Have a look at the new error messages